File: typescript/test/split.test.ts
import { containsDelimiter, splitBalanced } from "../src/parse";
import { describe, it, expect } from "vitest";

describe("containsDelimiter", () => {
    it("returns true when delimiter is present", () => {
        expect(containsDelimiter("a, b, c", ",")).toBe(true);
    });

    it("returns false when delimiter is absent", () => {
        expect(containsDelimiter("no commas here", ",")).toBe(false);
    });
});

describe("splitBalanced", () => {
    it("no_delimiter", () => {
        expect(splitBalanced("no commas here", ",")).toEqual(["no commas here"]);
    });

    it("basic_split", () => {
        expect(splitBalanced("a, b, c", ",")).toEqual(["a", " b", " c"]);
    });

    it("nested_parens", () => {
        expect(splitBalanced(":is(.a, .b), .c", ",")).toEqual([
            ":is(.a, .b)",
            " .c",
        ]);
    });

    it("nested_brackets", () => {
        expect(splitBalanced('[attr="x,y"], .z', ",")).toEqual([
            '[attr="x,y"]',
            " .z",
        ]);
    });

    it("quoted_strings (double quotes)", () => {
        expect(splitBalanced('"a,b", c', ",")).toEqual(['"a,b"', " c"]);
    });

    it("quoted_strings (single quotes)", () => {
        expect(splitBalanced("'a,b', c", ",")).toEqual(["'a,b'", " c"]);
    });

    it("deep_nesting", () => {
        expect(splitBalanced(":is(:not(.a, .b), .c), .d", ",")).toEqual([
            ":is(:not(.a, .b), .c)",
            " .d",
        ]);
    });

    it("empty_segments", () => {
        expect(splitBalanced(",a,,b,", ",")).toEqual(["", "a", "", "b", ""]);
    });

    it("single_char_delimiter", () => {
        expect(splitBalanced(",", ",")).toEqual(["", ""]);
    });

    it("empty_input", () => {
        expect(splitBalanced("", ",")).toEqual([""]);
    });

    it("mixed_nesting_and_quotes", () => {
        expect(
            splitBalanced(
                ':is(.a, .b), [data-x="1,2"], \'hello, world\'',
                ","
            )
        ).toEqual([":is(.a, .b)", ' [data-x="1,2"]', " 'hello, world'"]);
    });
});

File: rust/parse_that/tests/split_test.rs
use parse_that::split_balanced;

#[test]
fn no_delimiter() {
    assert_eq!(
        split_balanced("no commas here", b','),
        vec!["no commas here"]
    );
}

#[test]
fn basic_split() {
    assert_eq!(split_balanced("a, b, c", b','), vec!["a", " b", " c"]);
}

#[test]
fn nested_parens() {
    assert_eq!(
        split_balanced(":is(.a, .b), .c", b','),
        vec![":is(.a, .b)", " .c"],
    );
}

#[test]
fn nested_brackets() {
    assert_eq!(
        split_balanced(r#"[attr="x,y"], .z"#, b','),
        vec![r#"[attr="x,y"]"#, " .z"],
    );
}

#[test]
fn quoted_strings() {
    assert_eq!(split_balanced(r#""a,b", c"#, b','), vec![r#""a,b""#, " c"],);
}

#[test]
fn single_quoted_strings() {
    assert_eq!(split_balanced("'a,b', c", b','), vec!["'a,b'", " c"],);
}

#[test]
fn deep_nesting() {
    assert_eq!(
        split_balanced(":is(:not(.a, .b), .c), .d", b','),
        vec![":is(:not(.a, .b), .c)", " .d"],
    );
}

#[test]
fn empty_segments() {
    assert_eq!(split_balanced(",a,,b,", b','), vec!["", "a", "", "b", ""]);
}

#[test]
fn single_char() {
    assert_eq!(split_balanced(",", b','), vec!["", ""]);
}

#[test]
fn empty_input() {
    assert_eq!(split_balanced("", b','), vec![""]);
}

#[test]
fn mixed_nesting_and_quotes() {
    // CSS-like: :is(.a, .b), [data-x="1,2"], 'hello, world'
    assert_eq!(
        split_balanced(r#":is(.a, .b), [data-x="1,2"], 'hello, world'"#, b','),
        vec![":is(.a, .b)", r#" [data-x="1,2"]"#, " 'hello, world'"],
    );
}
